### ISSUE TITLE
[Feature] Mouse wheel or page.scroll

### ISSUE THREAD

**Issue by jperl** [microsoft/playwright#1115]
It would be nice to have a cross browser way to test scrolling. 

Ex. via CDP

```
await client.send("Input.dispatchMouseEvent", { type: "mouseWheel", deltaX: 0, deltaY: 500, x: 0, y: 0 });
```

For an auto-wait API, you could keep trying to scroll to a point and timeout if it cannot reach it.

**Comment by dgozman** [comment]
Could you share more about scenario? Are there `scroll` handlers? Why not use `element.scrollIntoViewIfNeeded`?

**Comment by jperl** [comment]
One use case is to test an infinite scroll component. Keep scrolling down in the x or y axis until the scrollLeft / scrollTop reaches a value. 

Right now I am doing this with [evaluate](https://github.com/qawolf/qawolf/blob/master/packages/web/src/scroll.ts) which works, I was just hoping to have a playwright API I could use.

**Comment by dgozman** [comment]
I think this evaluate script is reasonable for now, because it matches the exact semantics you care about. If we were to introduce scrolling, it would be more like user interaction as you proposed (either a mouse wheel scroll or touchscreen/touchpad scroll gesture), and it's unclear how to "wait for completion" in this case.

For infinite lists, perhaps `(await page.$('list-footer')).scrollIntoViewIfNeeded()` could just work?

**Comment by jperl** [comment]
Sounds good, I have workarounds but will leave this issue open for a potential future API. 

Right now I am using [`Input.dispatchMouseEvent`](https://github.com/qawolf/qawolf/blob/master/packages/web-tests/tests/Recorder.test.ts#L100-L106) when I need to test the user behavior. It would be nice to have a cross-browser playwright API like that. 

The way the auto-waiting could work is that it would resolve after the wheel event was completed. If the wheel event could somehow not be completed (maybe if the page could not scroll that amount) it would throw an error. 

The problem with using Input.dispatchMouseEvent now is that I do not know when it is finished, which is why this playwright api would be nice.

**Comment by alexsorinsky** [comment]
Definitely **mouseWheel** is very important feature. we have to have some well-known Official API way to handle with infinite scrolls when content will be downloaded dynamically when user reaches the end of the page

plus some quick-nav methods like
* scrollPageToBottom
* scrollPageToTop

https://codecept.io/helpers/Playwright/#scrollpagetobottom

+1 for API-native **mouseWheel**

**Comment by OlgaKuksa** [comment]
+1 for API-native mouseWheel or scroll to bottom
sample case - page.click() doesn't scroll to the element in the bottom of react-virtualized list of elements

**Comment by alexsorinsky** [comment]
it should be really awesome and i've been waiting for this feature very hard
but by the way could anyone of developers explain us what does "Native input events for mouse and keyboard" mean exactly?

are they really OS-native (putting mouse & keyboard like events into Browser's Window Procedure/Native Browser Event loop) or just emulated with some JS ?

ps. just cant found more appropriate place for this Q such as i'm asking playwright's devs personally to know exactly what "Native input" means here.

thx

**Comment by Tyriar** [comment]
I'd like this for xterm.js, I'm currently figuring out how to update our playwright private API usage that ensures wheel events get correctly serialized depending on the terminal mouse mode:

https://github.com/xtermjs/xterm.js/blob/4e18a78c02d0bcb5b97eeeb741d5e3c0b996cdc3/test/api/MouseTracking.api.ts#L77-L98

**Comment by colemars** [comment]
> I think this evaluate script is reasonable for now, because it matches the exact semantics you care about. If we were to introduce scrolling, it would be more like user interaction as you proposed (either a mouse wheel scroll or touchscreen/touchpad scroll gesture), and it's unclear how to "wait for completion" in this case.

Perhaps a "Scroll until some element appears" behavior? That would likely be a solid solution for infinite lists.

**Comment by DJ-Glock** [comment]
I have another case when mouse wheel is required - zooming. I have a webpage where zooming in/out works only with Ctrl+Mouse Wheel. So there is not way to do it with scrollIntoViewIfNeeded :)

**Comment by alexsorinsky** [comment]
well, its really logn to wait for this feature to be implemented, so i found much better alternative - [ui.vision](https://ui.vision/) 'coz of this - https://ui.vision/rpa/docs/xclick

look how this behavior been implemented there - ui.vision is just exactly what all of us need!

**Comment by yairEO** [comment]
I also need to test mouse wheel. I have `<input type="range">` where the mouse wheel (over the input element) controls the value, and this is extremely important in the implementation of my component, where normal dragging the thumb is not sensitive enough, so wheel movement is sometimes prefered. 

I would love to see support for simulating `wheel` event

**Comment by yifeikong** [comment]
Another workaround of `page.scroll` is to simply use `page.keyboard.press("PageDown")`.

**Comment by devniel** [comment]
+1 for this feature, especially for tables, when scrolling horizontally.

**Comment by etale-cohomology** [comment]
Without this feature how are we supposed to load dynamic content from websites that load content by scrolling down

**Comment by sergioariveros** [comment]
+1 We also need this to navigate inside a carousel

**Comment by racingcow** [comment]
+1 We need it to test zoom in/out of a map component.

**Comment by mxschmitt** [comment]
Closing since landed in #8690
